### ISSUE TITLE
[RFC] Scripts should use `cwd` instead of `work_dir`

### ISSUE THREAD

**Issue by mitchhentges** [mozilla-releng/scriptworker#290]
When scriptworker performs each build, it prepares a directory in which the script can work - a "work dir" (e.g.: a location to clone a repository to, or a holding place for tokens and nonces before signing).

Right now, we inform scripts of their "work dir" by passing a parameter as part of their config called `work_dir`.

**Suggestion:**
Instead, perhaps we should use the OS-level abstraction of the "current working directory", and remove the concept of an explicit `work_dir` configuration property

-----

**Benefits:**
* Simpler scripts - we don't need to merge `work_dir` with our relative paths - we can just use the relative paths directly
* Less footguns for new developers: "Why is my script failing due to not having file system permissions? 
Oh, because the current working directory is somewhere read-only, and I need to use `work_dir`"

-----

**Potential Concerns:**

Q: Doesn't this increase complexity, since, prior to running each script, we need to make sure it's running in `work_dir`, then go back to our old working directory afterwards?
A: Arguably, yes, `scriptworker` itself needs more logic. However right now, we're pushing concerns to each `script`, which means we're solving the issue `$n` times, instead of once at the source.

Q: Is using the current working directory less understandable than `work_dir` since it's less explicit? Will developers be confused how each `script` "just knows" where to create its files?
A: Since the "current working directory" abstraction is so common in the "system" domain, I don't think leveraging it will cause confusion. When I run `cat` or `jarsigner`, they operate in the current working directory implicitly, yet their behaviour is acceptable.
A potential parallel argument would be: I have an HTTP server. To access each endpoint, you _explicitly_ send a header: `Endpoint: /login`. Is this more maintainable than having the endpoint as part of the URL (`http://server/login`), which is a more-conventional method of conveying the concept of "endpoint" in HTTP?

-----

**Additional Notes:**

Other CI systems (e.g. [Gitlab Runner](https://docs.gitlab.com/runner/)) have a similar concept to `scriptworker` of having a single directory in which to run multiple sequential builds, cleaning the directory between runs. When Gitlab Runner invokes scripts for each new build, it sets up the "current working directory" as described in this RFC

**Comment by tomprince** [comment]
I'm not sure we should be using the current directory, but I definitely agree we shouldn't be hard-coding the work-dir in the static configuration.

**Comment by mitchhentges** [comment]
What concerns do you see with using the current directory?

**Comment by tomprince** [comment]
> What concerns do you see with using the current directory?

I don't have any specific concerns, and am not necessarily opposed to using it; I haven't given it a lot of  thought[1].  I mentioned it to make clear that, while I'm undecided on the proposed solution, while I strongly agree about the underlying issue.

[1] My gut reaction is to prefer being explicit, rather than depending on implicit global state.

**Comment by escapewindow** [comment]
Tl;dr: I don't necessarily agree, but I'm ok with it as long as we have the option of being explicit in production.

> Simpler scripts - we don't need to merge work_dir with our relative paths - we can just use the relative paths directly

If we maintain ability to use an explicit `work_dir`, we lose this benefit.

> Less footguns for new developers: "Why is my script failing due to not having file system permissions?

I don't see this helping new developers that much since we also require that the `cot/` directory exists inside of `work_dir`, so devs can't just run the script from any old directory; they need to populate the cot artifacts before they can run. If they're already populating a directory to use it, whatever script or documentation they use to do so could also specify editing their `script_config.json` to point the `work_dir` to the appropriate directory, or there could be local automation to do this.

However, I think it's fine to have a default of cwd for developers. I want this to be set explicitly for production, so ideally the default would be `...` for prod. I'm not sure the best way to do this: have a prod/dev switch? allow for python config files so we can use `os.getcwd()` in `script_config.py` ? Just have a default of `os.getcwd()` and make sure all prod configs have an explicit entry set?

> What concerns do you see with using the current directory?

a. I have muscle memory of launching the scripts from any directory, as long as their config is correct. Others may have the same. We're not removing a dev footgun; we're pointing it at other people's feet.
b. Similarly, on production instances, sometimes I launch the script without scriptworker, more footguns.
c. Even if I remember that I need to change directories to launch the script, instead of a simple single command with abspaths, I have to pushd/run script/popd every time I want to launch the script. I'm often cd'ed into a libpath or elsewhere, so just parking in `work_dir` is not a great option. Could also be switching screen/tmux panes, but it's still additional hassle when I'm debugging something.
d. On `signing-linux-dev1` we have many many `work_dir`s set up for testing various configs as well as load testing across dozens of script runs. To make this work, we set up the config properly for each separate `work_dir`, and once that's done, we know it'll behave properly across every run. We have separate runtime bustage possibilities if we rely on cwd, since operator error across any of the runs could lead to running in the wrong cwd while still using the proper script and config.

Also, [explicit is better than implicit](https://legacy.python.org/dev/peps/pep-0020/).

So at best I see this as a lateral change. At worst I see this introducing many more footguns than it fixes, while being less pythonic, less explicit, and solvable by other means.

**Comment by escapewindow** [comment]
(Hm, I went from the tl;dr being ok with it, to the final statement being mostly against it.)

**Comment by escapewindow** [comment]
I think this can be summed up as, making the `work_dir` specification a once-per-directory setup footgun a once-per-script-run footgun.

**Comment by tomprince** [comment]
An alternative, would be to pass the `work_dir` as an argument from scriptworker to the individual scripts, which would make the testing case easier.

**Comment by escapewindow** [comment]
> An alternative, would be to pass the `work_dir` as an argument from scriptworker to the individual scripts, which would make the testing case easier.

Could do. Would we have a default of `os.getcwd()` or `./work`, or make that arg required?

**Comment by tomprince** [comment]
> > An alternative, would be to pass the `work_dir` as an argument from scriptworker to the individual scripts, which would make the testing case easier.
> 
> Could do. Would we have a default of `os.getcwd()` or `./work`, or make that arg required?

My inclination would be to make it required (explicit rather than implicit), but it would be easy enough to change that in the future.

**Comment by escapewindow** [comment]
Hm, maybe `work_dir` in the config file, and allow for optionally overriding via commandline argument.

**Comment by mitchhentges** [comment]
Interesting points - I'm trying to think of another piece of software (that isn't a daemon) that has its working directory set in config, and I'm coming up a bit short. Maybe that's because we use our `scriptworker` `scripts` as _if_ they're daemons? :thinking: 
Either way, this is definitely valueable context.

Related, but if the net change was just to _add_ functionality (leave `work_dir` in config, _add command-line arg for override_) I'd prefer to defer that work until we have a need for that new functionality.

**Comment by escapewindow** [comment]
Good to know. I think ideally, the scripts would be much more script-like. Requiring existing artifacts in `cot/` and `task.json` in `work_dir/` definitely make them different than most other scripts. I'm not sure how best to deal with those to make the scripts more useful locally (an explosion of commandline arguments? a pregenerated manifest? wouldn't that manifest be similar to specifying `work_dir` when testing locally?) without adding too much complexity. But I'm definitely interested in making local testing easier, as long as it doesn't affect prod use cases worse.

**Comment by mitchhentges** [comment]
This isn't a pressing issue, and I misunderstood the context in which our `*scripts` are used. Closing for now
